core/matrix/hyb.cpp

```diff
@@ -0,0 +1,249 @@
+{
+    size_type nnz = 0;
+    IndexType current_row = 0;
+    auto row_nnz_val = row_nnz.get_data();
+    for (size_type i = 0; i < row_nnz.get_num_elems(); i++) {
+        row_nnz_val[i] = zero<size_type>();
+    }
+    for (const auto &elem : data.nonzeros) {
+        if (elem.row != current_row) {
+            row_nnz_val[current_row] = nnz;
+            current_row = elem.row;
+            nnz = 0;
+        }
+        nnz += (elem.value != zero<ValueType>());
+    }
+    row_nnz_val[current_row] = nnz;
+    return;
+}
+
+
+}  // namespace
+
+
+template <typename ValueType, typename IndexType>
+void Hyb<ValueType, IndexType>::apply_impl(const LinOp *b, LinOp *x) const
+{
+    using Dense = Dense<ValueType>;
+    this->get_executor()->run(
+        TemplatedOperation<ValueType, IndexType>::make_spmv_operation(
+            this, as<Dense>(b), as<Dense>(x)));
+}
+
+
+template <typename ValueType, typename IndexType>
+void Hyb<ValueType, IndexType>::apply_impl(const LinOp *alpha, const LinOp *b,
+                                           const LinOp *beta, LinOp *x) const
+{
+    using Dense = Dense<ValueType>;
+    this->get_executor()->run(
+        TemplatedOperation<ValueType, IndexType>::make_advanced_spmv_operation(
+            as<Dense>(alpha), this, as<Dense>(b), as<Dense>(beta),
+            as<Dense>(x)));
+}
+
+
+template <typename ValueType, typename IndexType>
+void Hyb<ValueType, IndexType>::convert_to(Dense<ValueType> *result) const
+{
+    auto exec = this->get_executor();
+    auto tmp = Dense<ValueType>::create(exec, this->get_size());
+    exec->run(TemplatedOperation<
+              ValueType, IndexType>::make_convert_to_dense_operation(tmp.get(),
+                                                                     this));
+    tmp->move_to(result);
+}
+
+
+template <typename ValueType, typename IndexType>
+void Hyb<ValueType, IndexType>::move_to(Dense<ValueType> *result)
+{
+    this->convert_to(result);
+}
+
+
+template <typename ValueType, typename IndexType>
+void Hyb<ValueType, IndexType>::read(const mat_data &data)
+{
+    Array<index_type> row_nnz(this->get_executor()->get_master(),
+                              data.size.num_rows);
+    get_each_row_nnz(data, row_nnz);
+    auto row_nnz_val = row_nnz.get_data();
+    std::sort(row_nnz_val, row_nnz_val + row_nnz.get_num_elems());
+
+    // get the limitation of columns of the ell part
+    index_type ell_lim = zero<index_type>();
+    if (method_ == partition::automatically) {
+        auto percent_pos = data.size.num_rows * 80 / 100;
+        ell_lim = row_nnz_val[percent_pos];
+    } else if (method_ == partition::percent) {
+        if (method_val_ == 100) {
+            ell_lim = row_nnz_val[row_nnz.get_num_elems() - 1];
+        } else {
+            auto percent_pos = data.size.num_rows * method_val_ / 100;
+            ell_lim = row_nnz_val[percent_pos];
+        }
+    } else if (method_ == partition::columns) {
+        ell_lim = method_val_;
+    }
```

Comment:
I also move the function for calculating the coo_lim into strategy to avoid calculating the nnz_per_row twice.
`strategy_->get_hybrid_limit(this->get_executor()->get_master(), data,
                                &ell_lim, &coo_lim);`